### PR TITLE
New object property to hide selection box.

### PR DESCRIPTION
**Pull request by Unarelith**
What the title says, it could be nice to have this feature for mobs and some item drop mods.

The new property is called `show_selection_box` but feel free to tell me if you want to change.

To test it, just run `/test_entity_nobox` in `minimal`.

screenshot-20181219153502

### CODE REVIEW

**Comment by nerzhul** [comment]
i agree.
As a complement, can you add a real use case in minimal/experimental ? Since 5.X dev cycle we like to have a real test case in minimal :)

**Comment by DS-Minetest** [comment]
> > There's a setting called `show_entity_selection_box`. Could it be changed to have three states, show all boxes, show no boxes and default?
> 
> This could be interesting for another PR.

Have you tested how your PR handles the setting currently?

**Comment by Quent42340** [comment]
No, but based on the code (isSettingEnabled && doShowSelectionBox) it should work as expected, which is showing selection box only when both the setting and the properties are set.

**Comment by SmallJoker** [comment]
Did I miss anything? This is already implemented:
https://github.com/minetest/minetest/blob/b7eb81fed922f7215cfff1068389fa85b2ee4ab2/doc/lua_api.txt#L5459-L5460

**Comment by Quent42340** [comment]
`pointable` disables punching completely when set to `false` iirc

**Comment by paramat** [comment]
Already possible for entities #3992
```
#    Show entity selection boxes
show_entity_selectionbox (Show entity selection boxes) bool true
```
I can't see any justification for this feature for nodes.
:-1:

**Comment by paramat** [comment]
The selectionbox for an entity is essential to show when you have selected it, it's essential for picking things up and essential for punching mobs. I can't see any use in completely disabling it.
If you temporarily don't want to see selectionboxes you can use F1. Or for all entities you can use the global setting.
I'm :-1: currently.

**Comment by Quent42340** [comment]
Please explain why the selectionbox is _essential_ for picking things and punching mobs?

Especially punching mobs, quote one game that shows a selection box for mobs.

**Comment by paramat** [comment]
Sorry i decided too quickly, removed my -1, neutral while i consider.

Yes, it's not essential, but it's very helpful for showing when you have actually selected the object, especially since the selectionbox can be a different shape than the object.
Mobs that are horizontally long cannot have long selectionboxes because boxes are axis-aligned. So the selectionbox is a very different shape to the mob.
So often you will think you're pointing at the object when you're not, and the opposite.
When there are several objects close together it's not always obvious which one is selected.
It's very often not obvious whether you are pointing at an object or a node behind it.

I don't care what other games do, i don't consider that a valid argument. Those are other games, maybe they're wrong? =) All that matters is what is best for MT.
It is very helpful to have visual feedback for when you are pointing at an object, instead of guessing. MT players are very used to this, a lack of a selectionbox for an object feels realy uncertain to me.

The selectionbox is often left set to be identical to the collisionbox so is usually useful for punching.

However, there may be good reason for the server to set the visibility of selectionboxes for certain objects.

What do you mean by 'item drop mods' and what's the use for those?

**Comment by Quent42340** [comment]
> I don't care what other games do, i don't consider that a valid argument. Those are other games, maybe they're wrong? =) All that matters is what is best for MT.

Well, that's why it wasn't the only argument. :)
And if every game that has mobs decided not to show selection boxes, it's for **immersion**.
But btw, the selection/collision box thing is a real engine issue...

> What do you mean by 'item drop mods' and what's the use for those?

Mods like [this one](https://github.com/minetest-mods/item_drop) for example, which drops items on dig instead of putting them directly into player inventory. In this case it's not necessary to show selection boxes on `__builtin:item`.

**Comment by DS-Minetest** [comment]
`item_drop` mods can use `pointable = false` to forbid clicking up items entirely.

As I wrote, it would be better to make the setting better so it has three states. Else there will always be some force against this PR because of fears of not seeing what entity is selected.
(Or maybe always show box in debug mode.)

**Comment by Quent42340** [comment]
> Or maybe always show box in debug mode.

This could be a good idea, indeed, but is it that important for players to see it?

**Comment by jastevenson303** [comment]
>>I don't care what other games do [...] All that matters is what is best for MT.

The first thing I do, well besides switching to ESDF and changing FOV and mouse sensitivity, is turn the selection boxes off for the entities and nodes.

A game creator may wish to specify that a particular object/item is to always show (or hide) its selection box, regardless of user preference.

And yes, now that player's collision box can be changed, it has produced some interesting results.  One I've found is turning a player's `phyisical' object property on.  Paramat says players "aren't supposed to be physical," but I don't really buy that.

**Comment by stujones11** [comment]
> regardless of user preference.

I can also see situations where this may be desirable but the question is, should a game creator have the right to disregard user preferences?

**Comment by Quent42340** [comment]
> I can also see situations where this may be desirable but the question is, should a game creator have the right to disregard user preferences?

Definitely! If a game creator/mod developer doesn't want players to see these boxes, he should be able to hide them.

Just imagine a game where immersion is very important: the creator will want to hide these boxes.

And I still don't understand what is so important about showing these boxes, sincerely. Most players already disable them anyway.

**Comment by PiezoU005F** [comment]
I agree that the best approach would be to have selection boxes be toggled with debug mode (whether or not to have this be privilege-dependent is up to anyone, but I don't see it as being very cheat-y)

**Comment by PiezoU005F** [comment]
I agree with DS-Minetest, this should be enabled/disabled with debug mode (whether or not to require the debug privilege is up to anyone, but if you ask me, this isn't very cheat-y, just aesthetically annoying).

File: src/object_properties.cpp [hunk 1]
```diff
@@ -167,4 +169,5 @@ void ObjectProperties::deSerialize(std::istream &is)
 	eye_height = readF1000(is);
 	zoom_fov = readF1000(is);
 	use_texture_alpha = readU8(is);
+	show_selection_box = readU8(is);
```
Comment (SmallJoker) [inline src/object_properties.cpp]: This may read garbage. Check the remaining stream length.

**Comment by paramat** [comment]
> Paramat says players "aren't supposed to be physical," but I don't really buy that.

That's just how MT implements players, it's not a matter of opinion, you can do what you want but weirdness may result.
> Most players already disable them anyway.

I don't think we know that for sure, the default setting seems more likely to be used.
Anyway, i'm neutral =)

**Comment by DS-Minetest** [comment]
I'm still for always showing the selection box in debug mode.

**Comment by LoneWolfHT** [comment]
If the selectionbox is hidden I would like to still have some way of knowing that I can interact with the entity/player

**Comment by ClobberXD** [comment]
> If the selectionbox is hidden I would like to still have some way of knowing that I can interact with an entity/player. Maybe by changing the pointer color

I think the whole point of hiding the selection box is to obfuscate the very fact that a certain object in the environment can be interacted with. This feature is useful for stuff like detective-style games, clues, hidden objects, camouflaged objects, etc.

**Comment by rubenwardy** [comment]
No response for other a month

**Comment by paramat** [comment]
Seems long neglected, i will notify the author and close for now, can always reopen.
@Unarelith

**Comment by Unarelith** [comment]
I can try to rework this PR if needed. The only thing that was left to do was to show the selection box all the time in debug mode, right?

**Comment by SmallJoker** [comment]
> The only thing that was left to do was to show the selection box all the time in debug mode, right?

Right.

**Comment by paramat** [comment]
This has an approval and support from at least 2 core devs, so is welcome. Will reopen PR on progress.

**Comment by Unarelith** [comment]
Ok great, I'll take care of that next week.
The only information I'll need though, is how to get this debug setting from my code.

**Comment by Unarelith** [comment]
@paramat I rebased it but this PR isn't updating, I don't know why.

Here is the correct commit: https://github.com/Unarelith/minetest/commit/dd775b2e3ee8d61b371cf4fa6613abbf9b37c9f7

Maybe try to reopen this one, and if it doesn't work I'll open another PR.

**Comment by paramat** [comment]
Github will not let me reopen this PR due to your branch being 'force pushed or recreated'.

**Comment by Unarelith** [comment]
@paramat #9255

**Pull request closed without merging**